storage/plans/iscsi/iscsi.go

```diff
@@ -0,0 +1,390 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package iscsi
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/retry"
+	"github.com/juju/utils/clock"
+	"github.com/juju/utils/exec"
+
+	"github.com/juju/juju/storage"
+	"github.com/juju/juju/storage/plans/common"
+)
+
+var logger = loggo.GetLogger("juju.storage.plans.iscsi")
+
+const (
+	// ISCSI_ERR_SESS_EXISTS is the error code open-iscsi returns if the
+	// target is already logged in
+	ISCSI_ERR_SESS_EXISTS = 15
+
+	// ISCSI_ERR_NO_OBJS_FOUND is the error code open-iscsi returns if
+	// no records/targets/sessions/portals are found to execute operation on
+	ISCSI_ERR_NO_OBJS_FOUND = 21
+)
+
+var (
+	sysfsBlock        = "/sys/block"
+	sysfsClass        = "/sys/class"
+	sysfsiSCSIHost    = "/sys/class/iscsi_host"
+	sysfsiSCSISession = "/sys/class/iscsi_session"
+
+	iscsiConfigFolder = "/etc/iscsi"
+)
+
+type iscsiPlan struct{}
+
+func NewiSCSIPlan() common.Plan {
+	return &iscsiPlan{}
+}
+
+func (i *iscsiPlan) AttachVolume(volumeInfo map[string]string) (storage.BlockDevice, error) {
+	plan, err := newiSCSIInfo(volumeInfo)
+	if err != nil {
+		return storage.BlockDevice{}, errors.Trace(err)
+	}
+	return plan.attach()
+}
+
+func (i *iscsiPlan) DetachVolume(volumeInfo map[string]string) error {
+	plan, err := newiSCSIInfo(volumeInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return plan.detach()
+}
+
+type iscsiConnectionInfo struct {
+	iqn        string
+	address    string
+	port       int
+	chapSecret string
+	chapUser   string
+}
+
+var runCommand = func(params []string) (*exec.ExecResponse, error) {
+	cmd := strings.Join(params, " ")
+	execParams := exec.RunParams{
+		Commands: cmd,
+	}
+	resp, err := exec.RunCommands(execParams)
+
+	return resp, err
+}
+
+func getHardwareInfo(name string) (storage.BlockDevice, error) {
```

Comment:
I think it would make sense to split this into 2 separate functions for testing purposes. One would actually call an external command, and then just return an io.Reader that you could use.
Then you can use a bytes.Buffer to test the function that parses the result, and easily feed that function test input so we can make sure it copes with any edge cases that we find in the wild.

Reply:
I think we can just get away with mocking out ```runCommand``` and returning whatever output we want?